src/celeritas/ext/GeantSurfacePhysicsLoader.cc

```diff
@@ -0,0 +1,469 @@
+ */
+bool total_prob_is_unity(inp::ReflectionForm const& rf)
+{
+    auto const& sl = rf.specular_lobe;
+    auto const& ss = rf.specular_spike;
+    auto const& bc = rf.backscatter;
+    auto const& dl = rf.diffuse_lobe;
+    auto const size = sl.x.size();
+    CELER_ASSERT(ss.x.size() == size && bc.x.size() == size
+                 && dl.x.size() == size);
+
+    auto prob_sum = [&](size_type index) -> real_type {
+        return sl.y[index] + ss.y[index] + bc.y[index] + dl.y[index];
+    };
+
+    for (auto i : range(sl.x.size()))
+    {
+        if (!soft_equal(real_type{1}, prob_sum(i)))
+        {
+            return false;
+        }
+    }
+    return true;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate diffuse lobe from the rest of the imported data.
+ *
+ * Since the total probability for all 4 properties is equal to one, the
+ * diffuse lobe can be calculated by subtracting the other three.
+ */
+inp::Grid calc_diffuse_lobe(inp::ReflectionForm const& refl_form)
+{
+    auto const& sl = refl_form.specular_lobe;
+    auto const& ss = refl_form.specular_spike;
+    auto const& bc = refl_form.backscatter;
+    auto const size = sl.x.size();
+    CELER_ASSERT(ss.x.size() == size && bc.x.size() == size);
+
+    inp::Grid result;
+    result.x = sl.x;
+    result.y.resize(size);
+    for (auto i : range(size))
+    {
+        // diffuse_lobe = 1 - specular_lobe - specular_spike - backscatter
+        result.y[i] = real_type{1} - sl.y[i] - ss.y[i] - bc.y[i];
+    }
+    CELER_ENSURE(result);
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+}  // namespace
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct empty.
+ */
+GeantSurfacePhysicsLoader::GeantSurfacePhysicsLoader() {}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Populate surface physics data for a given \c SurfaceId .
+ */
+void GeantSurfacePhysicsLoader::operator()(SurfaceId sid,
+                                           inp::SurfacePhysics& result)
+{
+    CELER_EXPECT(sid);
+    detail::GeantSurfacePhysicsHelper helper(sid);
+    try
+    {
+        this->insert_reflectivity(helper, result);
+        this->insert_roughness(helper, result);
+        this->insert_interaction(helper, result);
+
+        // Ensure that data is compatible with selected model
+        this->validate_model(helper, result);
+    }
+    catch (RuntimeError const& e)
+    {
+        auto const& surf = helper.surface();
+        CELER_LOG(error) << "failed to convert surface " << surf.GetName()
+                         << " (id " << sid.unchecked_get() << ") with model "
+                         << to_cstring(surf.GetModel()) << " and finish "
+                         << to_cstring(surf.GetFinish()) << ": "
+                         << e.details().which << ", " << e.details().what;
+    }
```

Comment:
I meant to log the error's context and then keep propagating the exception so the code doesn't keep going
```suggestion
        throw;
    }
```